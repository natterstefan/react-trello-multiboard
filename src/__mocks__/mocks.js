import { pick } from 'lodash'

// This file contains mocked responses from the data/trello API.
export const mockExampleBoardConfig = {
  board: 'hello-world',
  estimates_with_round_brackets: true,
  estimates_with_square_brackets: true,
  lists: /#upcoming/,
  shortcut: 'hw',
}

export const mockEampleUser1 = {
  avatarHash: '123456789',
  fullName: 'Example Username',
  id: 'member-1',
  initials: 'EU',
  username: 'exampleusername',
}

export const mockEampleUser2 = {
  avatarHash: '987654321',
  fullName: 'Another User',
  id: 'member-2',
  initials: 'AU',
  username: 'anotheruser',
}

export const mockEampleUser3 = {
  avatarHash: null,
  fullName: 'Third User',
  id: 'member-3',
  initials: 'TU',
  username: 'thirduser',
}

export const mockExampleMemberResponse = username => {
  switch (username) {
    case 'anotheruser':
      return mockEampleUser2
    case 'thirduser':
      return mockEampleUser3
    default:
      return mockEampleUser1
  }
}

export const mockExampleBoardResponse = { name: 'hello-world', id: 'board-1' }

// most likely it should be an array with two lists (one matching and one not)
export const mockExampleListResponse = {
  id: 'list-123',
  idBoard: 'board-1',
  name: 'Test List Name #upcoming',
}

export const mockExampleCardsResponse = {
  id: 'card-1',
  name: 'Test Card Name',
  idBoard: 'board-1',
  idMembers: [mockEampleUser1.id, mockEampleUser2.id],
  idList: 'list-1',
  shortUrl: 'https://trello.com/c/abcdefgh',
  members: [
    pick(mockEampleUser1, ['id', 'username', 'fullName']),
    pick(mockEampleUser2, ['id', 'username', 'fullName']),
  ],
}
