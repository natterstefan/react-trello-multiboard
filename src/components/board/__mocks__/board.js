import {
  mockExampleUser1,
  mockExampleUser2,
  mockExampleBoardResponse,
  mockExampleListResponse,
} from '../../../__mocks__/mocks'

export const props = {
  board: {
    id: 'board-1',
  },
}

export const storeStateMock = {
  boards: {
    data: [
      {
        board: mockExampleBoardResponse,
      },
      {
        board: {
          id: 'board-2',
          name: 'another-example-board',
        },
      },
    ],
  },
  lists: {
    'board-1': {
      isLoading: false,
      data: [mockExampleListResponse],
      error: null,
    },
  },
  info: {
    version: 1,
  },
  app: {
    memberToggle: {
      togglePreferred: false,
      togglePreferredMember: null,
    },
  },
  members: {
    members: [mockExampleUser1, mockExampleUser2],
  },
}
