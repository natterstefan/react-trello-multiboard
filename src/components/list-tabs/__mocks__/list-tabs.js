export const storeStateMock = {
  app: {
    config: {
      lists: ['list-1', 'list-2'],
    },
    listToggle: {
      toggleList: 'list-1',
    },
  },
}
