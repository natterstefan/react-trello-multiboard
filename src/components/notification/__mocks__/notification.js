export const props = { message: 'some notification' }

export const storeStateMock = {
  boards: {
    isLoading: false,
    error: null,
  },
  lists: {
    'board-1': {
      isLoading: false,
      error: null,
    },
  },
  members: {
    isLoading: false,
    error: null,
  },
}

export const storeWithBoardError = {
  boards: {
    isLoading: false,
    error: {
      status: 401,
      responseText: 'example response',
    },
  },
}

export const storeWithAnyError = {
  user: {
    error: { message: 'some error user message ' },
  },
}
