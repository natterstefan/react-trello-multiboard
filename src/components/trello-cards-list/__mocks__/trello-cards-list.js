import {
  mockExampleCardsResponse,
  mockExampleBoardConfig,
  mockExampleListResponse,
} from '../../../__mocks__/mocks'

export const props = {
  list: mockExampleListResponse,
  config: mockExampleBoardConfig,
}

export const storeStateMock = {
  cards: {
    'list-1': {
      data: [
        {
          card: mockExampleCardsResponse,
          config: mockExampleBoardConfig,
        },
      ],
      isLoading: true,
      error: null,
    },
  },
}
