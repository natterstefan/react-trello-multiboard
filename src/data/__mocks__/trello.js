// JEST - How to mock modules
//
// either we mock it in each test seperately
// - >>> https://github.com/facebook/jest/issues/2582#issuecomment-346886018
// - https://stackoverflow.com/a/40465435/1238150
//
// or provide a solution for every test in the __mocks__
// - https://facebook.github.io/jest/docs/en/manual-mocks.html
import {
  mockExampleMemberResponse,
  mockExampleBoardResponse,
  mockExampleListResponse,
  mockExampleCardsResponse,
} from '../../__mocks__/mocks'

// NOTE: use mockResolvedValue instead
export const authenticateUser = jest.fn().mockResolvedValue({})
export const getBoard = jest.fn().mockResolvedValue(mockExampleBoardResponse)
export const getMeBoards = jest.fn().mockResolvedValue([mockExampleBoardResponse])
export const getLists = jest.fn().mockResolvedValue([mockExampleListResponse])
export const getCards = jest.fn().mockResolvedValue([mockExampleCardsResponse])
export const getMember = jest.fn(username => Promise.resolve(mockExampleMemberResponse(username)))
export const isTrelloAvailable = jest.fn().mockReturnValue(true)

export default {
  authenticateUser,
  getBoard,
  getMeBoards,
  getLists,
  getCards,
  getMember,
  isTrelloAvailable,
}
