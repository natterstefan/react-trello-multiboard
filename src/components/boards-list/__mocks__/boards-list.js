import { mockExampleUser1, mockExampleBoardResponse } from '../../../__mocks__/mocks'

export const storeStateMock = {
  boards: {
    data: [{ board: mockExampleBoardResponse }],
    error: null,
    isLoading: true,
  },
  app: {
    memberToggle: {
      togglePreferred: false,
      togglePreferredMember: null,
    },
  },
  members: { members: [mockExampleUser1] },
}
