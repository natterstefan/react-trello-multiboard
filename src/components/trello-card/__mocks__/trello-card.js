import { mockEampleUser1, mockEampleUser2 } from '../../../__mocks__/mocks'

export const props = {
  addEstimations: jest.fn(),
  boardName: 'example-board',
  card: {
    badges: {
      description: false,
      comments: 1,
      attachments: 1,
      checkItems: 2,
      checkItemsChecked: 1,
    },
    id: 'id-1',
    idBoard: 'board-1',
    idMembers: ['member-1', 'member-2'],
    members: [mockEampleUser1, mockEampleUser2],
    name: 'example-name',
    shortUrl: 'https://trello.com/id-1',
  },
  config: {
    estimates_with_round_brackets: false,
    estimates_with_square_brackets: false,
  },
  isHidden: false,
  listName: 'example-list',
}

export const storeStateMock = {
  boards: {
    data: [
      {
        board: {
          // will be added to the TrelloCard (because it matches the idBoards above)
          id: 'board-1',
          name: 'example-board',
        },
      },
      {
        board: {
          id: 'board-2',
          name: 'another-example-board',
        },
      },
    ],
  },
  app: {
    memberToggle: {
      togglePreferred: false,
      togglePreferredMember: null,
    },
  },
  members: {
    members: [],
  },
}

export const storeStateMockPreferred = {
  app: {
    memberToggle: {
      togglePreferred: true,
      togglePreferredMember: null,
    },
  },
  members: {
    members: [{ id: 'member-1' }, { id: 'member-2' }],
  },
}

export const storeStateMockPreferredMember = {
  app: {
    memberToggle: {
      togglePreferred: true,
      togglePreferredMember: 'member-1',
    },
  },
  members: {
    members: [{ id: 'member-1' }, { id: 'member-2' }],
  },
}

export const storeStateMockPreferredMemberInvalid = {
  app: {
    memberToggle: {
      togglePreferred: true,
      togglePreferredMember: 'member-3',
    },
  },
  members: {
    members: [{ id: 'member-1' }, { id: 'member-2' }],
  },
}
