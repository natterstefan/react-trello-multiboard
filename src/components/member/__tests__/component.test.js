import React from 'react'
import { shallow } from 'enzyme'
import { merge } from 'lodash'

import LinearProgress from '@material-ui/core/LinearProgress'
import Card from '@material-ui/core/Card'
import Typography from '@material-ui/core/Typography'

import { mockEampleUser1 } from '../../../__mocks__/mocks'

import Member from '../component'
import UserAvatar from '../user-avatar'

const props = {
  doTogglePreferredMember: jest.fn(),
  error: null,
  estimations: {
    consumed: 0,
    estimated: 0,
  },
  isActive: false,
  isLoading: false,
  member: mockEampleUser1,
}

describe('Component/Member', () => {
  test('should render without throwing an error', () => {
    expect(shallow(<Member {...props} />).dive()).toMatchSnapshot()
  })

  test('should render an UserAvatar per Member', () => {
    const wrapper = shallow(<Member {...props} />).dive()
    expect(wrapper.find(UserAvatar).length).toEqual(1)
  })

  test('should render a LinearProgress when component is waiting for loading data to finish', () => {
    const wrapper = shallow(<Member {...props} isLoading />).dive()
    expect(wrapper.find(LinearProgress).length).toEqual(1)
  })

  test('should render an empty span when an error occured', () => {
    const wrapper = shallow(<Member {...props} error="sample error" />).dive()
    expect(wrapper.html()).toEqual('<span></span>')
  })

  test('should change the styles of the Card and Typography when member isActive', () => {
    const wrapper = shallow(<Member {...props} isActive />).dive()
    expect(wrapper.find(Card).prop('style')).toMatchObject({ backgroundColor: '#3f51b5' })
    expect(wrapper.find(Typography).prop('style')).toMatchObject({ color: '#fff' })
  })

  test('should render no estimations if they are not available', () => {
    const newProps = merge({}, props, {
      estimations: null,
    })
    const wrapper = shallow(<Member {...newProps} />).dive()
    expect(wrapper.find('.member_card_member-1--estimations').length).toEqual(0)
  })

  test('should render estimations.consumed=0 when they are not available', () => {
    const newProps = merge({}, props, {
      estimations: {
        estimated: 2,
      },
    })
    const wrapper = shallow(<Member {...newProps} />).dive()
    const expectedSpan = '<span class="member_card_member-1--estimations">(2)[0]</span>'
    expect(wrapper.find('.member_card_member-1--estimations').html()).toEqual(expectedSpan)
  })

  test('should invoke doTogglePreferredMember when the Card is clicked', () => {
    const doTogglePreferredMember = jest.fn()
    const wrapper = shallow(
      <Member {...props} doTogglePreferredMember={doTogglePreferredMember} />,
    ).dive()
    wrapper.find('.member_card_member-1').simulate('click')
    expect(doTogglePreferredMember).toHaveBeenCalledTimes(1)
  })
})
