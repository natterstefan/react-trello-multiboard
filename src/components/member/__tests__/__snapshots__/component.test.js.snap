// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component/Member should render without throwing an error 1`] = `
<Card
  className="member_card_member-1"
  onClick={[Function]}
  raised={false}
  style={
    Object {
      "backgroundColor": undefined,
    }
  }
>
  <WithStyles(CardContent)
    className="Member-root-1"
  >
    <styled.div>
      <WithStyles(UserAvatar)
        member={
          Object {
            "avatarHash": "123456789",
            "fullName": "Example Username",
            "id": "member-1",
            "initials": "EU",
            "username": "exampleusername",
          }
        }
      />
      <WithStyles(Typography)
        style={
          Object {
            "color": undefined,
          }
        }
      >
        <span
          className="member_card_member-1--estimations"
        >
          (
          0
          )[
          0
          ]
        </span>
      </WithStyles(Typography)>
    </styled.div>
  </WithStyles(CardContent)>
</Card>
`;
